### ISSUE TITLE
OnSignInCompleted not triggered for android when custom login url is used for <OktaDomain>

### ISSUE THREAD

**Issue by PBhuva** [okta/okta-oidc-xamarin#44]
Hello Team,

We are trying out the stable version of Okta SDK i.e Okta.Xamarin(1.0.0).
We have observed one issue with Android. When we use custom login URL for OktaDomain configuration then once login is completed ,only browser gets closed but OnSignInCompleted event is not triggered.

## Current behavior
1. Use **custom login** URL for OktaDomain in OktaConfig.xml in Android
2. Call await OktaContext.Current.SignInAsync(); for login method
3. Okta custom login page will be displayed in the browser
4. Enter the credentials
5. After login is completed,browser gets closed but OnSignInCompleted is not triggered.Also the control does not go in catch block.

## Expected behavior
1. Use **custom login** URL for OktaDomain in OktaConfig.xml in Android
2. Call await OktaContext.Current.SignInAsync(); for login method
3. Okta custom login page will be displayed in the browser
4. Enter the credentials
5. After login is completed,browser gets closed and OnSignInCompleted should get triggered and will return the tokens.

## Environment

- OS:
     - Android
- Other:
     - Visual Studio for Mac version 8.9.7(build 8)

**Comment by bryanapellanes-okta** [comment]
@PBhuva,
To help reproduce what you are seeing can you share minimal code that illustrates the problem that occurs?

**Comment by PBhuva** [comment]
Hi @BryanApellanes,
I have uploaded a sample POC at https://github.com/PBhuva/CustomLoginURLSample 
Please replace the configurations with actual ones and use custom domain URL for OktaDomain in OktaConfig.xml to reproduce the issue.

**Comment by bryanapellanes-okta** [comment]
After review of https://developer.okta.com/docs/guides/custom-url-domain/enable-the-custom-domain/, I believe the issue is that the application is not set up to properly receive the redirects from the custom domain.  I'm reviewing this further to determine best approach to solution.  This may just need an update to the setup instructions.

**Comment by bryanapellanes-okta** [comment]
@PBhuva,
Can you confirm that this code executes when NOT using a custom login url? https://github.com/PBhuva/CustomLoginURLSample/blob/main/CustomLoginUrlPOC/CustomLoginUrlPOC/MainPageViewModel.cs#L29

**Comment by bryanapellanes-okta** [comment]
@PBhuva,
One thing to note, is that this line: https://github.com/PBhuva/CustomLoginURLSample/blob/main/CustomLoginUrlPOC/CustomLoginUrlPOC/MainPageViewModel.cs#L26
should come before this line: https://github.com/PBhuva/CustomLoginURLSample/blob/main/CustomLoginUrlPOC/CustomLoginUrlPOC/MainPageViewModel.cs#L22
This is to ensure that the event subscription is already in place before the sign-in process completes; the likelihood of this becoming a problem is small, but it's possible that sign-in completes before the signincompleted handler is attached.

However, this may or may not be the source of the issue you are experiencing; continuing to investigate.

**Comment by bryanapellanes-okta** [comment]
@PBhuva,
To help me reproduce the behavior you see, please confirm what device/emulator version you are using to run your POC.  Thanks!

**Comment by bryanapellanes-okta** [comment]
I've taken your sample code and updated the package references and the SignInCompleted event is triggered as expected, see: https://github.com/PBhuva/CustomLoginURLSample/pull/1

**Comment by PBhuva** [comment]
@bryanapellanes-okta  Thanks you for updates. I did tried out running the sample code on Android device but the issue is still the same for me.
Can you confirm if you tried it with **Custom** login url?
Also I tried replacing the okta configurations in the Okta Sample (https://github.com/okta/okta-oidc-xamarin) but facing the same issue.This issue exists only on Android and not on iOS.

**Comment by bryanapellanes-okta** [comment]
@PBhuva see response above.

**Comment by PBhuva** [comment]
@bryanapellanes-okta Just to provide additional information. If we use issuer URL for <OktaDomain> then SignInCompleted event is triggered but if we use custom login URL then the event is not triggered. For both the cases, configurations on the app side are same.

**Comment by bryanapellanes-okta** [comment]
> @bryanapellanes-okta Just to provide additional information. If we use issuer URL for then SignInCompleted event is triggered but if we use custom login URL then the event is not triggered. For both the cases, configurations on the app side are same.

@PBhuva  Please describe the difference between your **issuer URL** and the **custom login URL**; perhaps I'm misunderstanding how to reproduce what you are seeing.  Just to confirm, when you say **custom login URL**, you are referring to a URL that was setup using the instructions found here: https://developer.okta.com/docs/guides/custom-url-domain/enable-the-custom-domain/, is that correct?

In this demo video, using your sample, the custom url is `bapellanes.sigmanetcorp.us`: https://user-images.githubusercontent.com/63638027/125646419-bf52058b-5c1b-4808-a3b2-632764c356b0.mp4

**Comment by bryanapellanes-okta** [comment]
@PBhuva To help me further, please provide the contents of your `OktaConfig.xml` in the Android project Assets folder (ensure you redact any private information).

**Comment by PBhuva** [comment]
@bryanapellanes-okta  Please find below details of OktaConfig.xml for Android
<?xml version="1.0" encoding="UTF-8" ?>
<Okta>
    <ClientId>{ClientId}</ClientId>
    <OktaDomain>{Custom domain url}</OktaDomain>
    <RedirectUri>{RedirectUri}</RedirectUri>
    <PostLogoutRedirectUri>{Postlogout redirect uri}</PostLogoutRedirectUri>
    <Scope>openid profile offline_access</Scope>
    <AuthorizationServerId>{authorization serverid other than default}</AuthorizationServerId>
</Okta>

**Comment by PBhuva** [comment]
@bryanapellanes-okta  Please find below content of OktaConfig.xml for Android in Assets folder

![Screenshot 2021-07-28 at 8 18 47 PM](https://user-images.githubusercontent.com/62088523/127343810-9b2f435c-1e0e-4120-a2ce-67d8107eeecb.png)

**Comment by bryanapellanes-okta** [comment]
@PBhuva Please provide me the value of the Redirect Uri specified in the `OktaConfig.xml` file.  This should be something in the format that Android expects.  The value specified must match the the combined `DataSchemes` and `DataPath` properties of the `IntentFilter` defined on your `LoginCallbackInterceptorActivity`; this value must also be specified in the application configuration on the Okta dashboard.  It should look similar to `com.okta.xamarin.android.login:/callback`

**Comment by PBhuva** [comment]
@bryanapellanes-okta value for redirect uri is `com.mmi.debtmanagement-mmi-poc-177:/callback` 
Also for `LoginCallbackInterceptorActivity` values for DataSchemes and DataPath are as follows-
 ```
DataSchemes = new[] { "com.mmi.debtmanagement-mmi-poc-177" },
 DataPath = "/callback"
```

**Comment by bryanapellanes-okta** [comment]
@PBhuva, 
Thanks for your response.  Please confirm that you also have a corresponding entry in your Okta dashboard under general settings for your application.  It should look similar to the following:

![ApplicationGeneratlSettings](https://user-images.githubusercontent.com/63638027/127656446-8e5df40f-d9f3-4539-8b71-e80ed6d174ce.png)

**Comment by PBhuva** [comment]
@bryanapellanes-okta Please check the 2nd url Sign-in redirect URIs **com.mmi.debtmanagement-mmi-poc-177:/callback**
![MicrosoftTeams-image](https://user-images.githubusercontent.com/62088523/127811430-b8d5c5b4-b54e-4dfc-8e93-a8fa65958d0a.png)

**Comment by bryanapellanes-okta** [comment]
@PBhuva,
Thanks for your response!  In order to reproduce the behavior you are seeing, I'd like to configure your POC project with your configuration values.  In order for me to log in, please create a test user for me to test the sign in process.

**Comment by PBhuva** [comment]
Hi @bryanapellanes-okta I have updated the POC project with our configurations and provided you the collaborator access.The test credentials are in the comment above OnLogin method (https://github.com/PBhuva/CustomLoginURLSample/blob/main/CustomLoginUrlPOC/CustomLoginUrlPOC/MainPageViewModel.cs).
Note: OktaDomain in OktaConfig.xml is set to custom login URL and the commented URL is the issuer URL which works fine.

**Comment by bryanapellanes-okta** [comment]
@PBhuva 
After extensive investigation into this issue I've determined the root cause.  The root certificate authority for the custom domain is not trusted by the android device so, after authentication completes, the Sdk attempts to exchange the authentication code for tokens but the call fails because the root CA is not trusted.  See this line of code, https://github.com/okta/okta-oidc-xamarin/blob/release-v3.0.0-dev/Okta.Xamarin/Okta.Xamarin/OidcClient.cs#L489 

This issue cannot be corrected using only Xamarin tools, related Xamarin issue here: https://github.com/xamarin/xamarin-android/issues/2176

In order to trust the CA one must use one of the options described here: https://developer.android.com/training/articles/security-config

However, there is no easy intuitive way to add the xml attribute `android:networkSecruityConfig="@xml/network_security_config"` to the application element of the AndroidManifest.xml file.  To accomplish this, one must first build the Xamarin Android project, then navigate in the filesystem to the location the Xamarin build tools place temporary files and manually edit the AndroidManifest.xml file directly.  In my case the build output went to ./obj/Debug/110/

I will modify the Okta Sdk to bring awareness to our developers when this issue arises, by adding an `AuthCodeTokenExchangeFailed ` event.

**Comment by bryanapellanes-okta** [comment]
This is addressed in release v3, see also #66
